### PR TITLE
Enable retrieval of assessment data

### PR DESCRIPTION
**Pull request by PhilDigitalJustice**

### CODE REVIEW

File: assessment-api/open-api-specification.yml [hunk 1]
```diff
@@ -3,12 +3,29 @@ info:
...
+  /assessments:
     get:
       tags:
         - 'assessment'
-      summary: 'Assessment'
-      operationId: 'assessment'
+      summary: 'get assessments'
+      operationId: 'getAssessments'
+      parameters:
+        - $ref: '#/components/parameters/assessmentParameter'
```
Comment (porritta) [inline assessment-api/open-api-specification.yml:13]: Should this not be assessmentTypeParameter?

File: assessment-api/open-api-specification.yml [hunk 2]
```diff
@@ -24,8 +41,54 @@ paths:
...
 
 components:
+  parameters:
+    assessmentParameter:
+      name: name
+      in: query
+      schema:
+        type: string
+        enum:
+          - 'billingAssessment'
```
Comment (porritta) [inline assessment-api/open-api-specification.yml:82]: Is it possible to drop the 'Assessment' bit from these enum values, and make them more standard?
AssessmentType.BILLING
AssessmentType.POA
AssessmentType.POA_PREPOP

File: assessment-api/open-api-specification.yml [hunk 3]
```diff
@@ -40,14 +103,19 @@ components:
...
         id:
           type: 'string'
-        owner_id:
+        provider_id:
           type: 'string'
-        target_id:
+        case_reference_number:
           type: 'string'
-        assessment:
+        name:
```
Comment (porritta) [inline assessment-api/open-api-specification.yml:110]: Is name actually type?

File: assessment-api/open-api-specification.yml [hunk 4]
```diff
@@ -62,4 +130,79 @@ components:
...
+          default: [ ]
+          items:
+            $ref: '#/components/schemas/assessmentRelationshipDetail'
+        prepopulated:
+          type: boolean
+      required:
+        - id
+        - entityId
+        - prepopulated
+    assessmentEntityTypeDetail:
```
Comment (porritta) [inline assessment-api/open-api-specification.yml:157]: I know it makes no difference really, but if you move assessmentEntityTypeDetail above assessmentEntityDetail then the definitions are in the right order.

File: assessment-api/open-api-specification.yml (hunk 3 above)
Comment (PhilDigitalJustice) [reply assessment-api/open-api-specification.yml:110]: well the object is called assessmentDetail 

'name' is the name of the assessment

Comment (PhilDigitalJustice) [reply assessment-api/open-api-specification.yml:110]: it also follows a similar structure to the other objects,  attributs have a name, entity have a name, makes sense that assessment does.

File: assessment-api/open-api-specification.yml (hunk 2 above)
Comment (PhilDigitalJustice) [reply assessment-api/open-api-specification.yml:82]: these are the existing values for the assessment names, as much as i would agree this is what OPA is currently using and looking up through the connector. Once we have amended the connector to use this as well, we could easily amend this

File: assessment-api/open-api-specification.yml (hunk 1 above)
Comment (PhilDigitalJustice) [reply assessment-api/open-api-specification.yml:13]: will amend

File: assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaAttribute.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+   * Indicates if the attribute is prepopulated.
+   */
+  @Column(name = "PREPOPULATED", nullable = false)
+  private boolean prepopulated;
+
+  /**
+   * Indicates if the attribute was asked in the session.
+   */
+  @Column(name = "ASKED")
+  private boolean asked = false;
```
Comment (porritta) [inline assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaAttribute.java:83]: No issues with primitive and assignment here? Might we be doing Example.of?

Comment (PhilDigitalJustice) [reply assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaAttribute.java:83]: In don't think we will be enabling any endpoints to do this anytime soon. Primarily as we'll want to keep the /assessments endpoint in a single transaction. I would imagine we might in future want to look for specific attributes on another endpoint. I'll amend it.

File: assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaAttribute.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+   * The type of inferencing used for the attribute.
+   */
+  @Column(name = "INFERENCING_TYPE")
+  private String inferencingType;
+
+  /**
+   * Indicates if the attribute is prepopulated.
+   */
+  @Column(name = "PREPOPULATED", nullable = false)
+  private boolean prepopulated;
```
Comment (porritta) [inline assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaAttribute.java:77]: Do we not need to specify the Converter,  eg @Convert(converter = NumericBooleanConverter.class)

File: assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaEntity.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  private OpaSession opaSession;
+
+  /**
+   * The identifier of the entity.
+   */
+  @Column(name = "ENTITY_ID")
+  private String entityId;
+
+  /**
+   * The type of the entity.
```
Comment (porritta) [inline assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaEntity.java:63]: wrong jdoc

File: assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaRelationship.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+   * The name of the relationship.
+   */
+  @Column(name = "NAME")
+  private String name;
+
+  /**
+   * Indicates if the relationship is prepopulated.
+   */
+  @Column(name = "PREPOPULATED")
+  private boolean prepopulated;
```
Comment (porritta) [inline assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaRelationship.java:72]: Converter

Comment (porritta) [reply assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaRelationship.java:72]: And should be Boolean, if we're making them all the same

**Review by porritta — CHANGES_REQUESTED** [review]
comments added

File: assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaEntity.java (hunk 7 above)
Comment (PhilDigitalJustice) [reply assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaEntity.java:63]: amended

File: assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaRelationship.java (hunk 8 above)
Comment (PhilDigitalJustice) [reply assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaRelationship.java:72]: amended

File: assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaAttribute.java (hunk 6 above)
Comment (PhilDigitalJustice) [reply assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaAttribute.java:77]: amended

Comment (porritta) [reply assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaAttribute.java:80]: Should be a Boolean too

File: assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaAttribute.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+   */
+  @Column(name = "PREPOPULATED", nullable = false)
+  @Convert(converter = NumericBooleanConverter.class)
+  private boolean prepopulated;
+
+  /**
+   * Indicates if the attribute was asked in the session.
+   */
+  @Column(name = "ASKED")
+  private Boolean asked = false;
```
Comment (porritta) [inline assessment-service/src/main/java/uk/gov/laa/ccms/caab/assessment/entity/OpaAttribute.java:86]: I missed this one - needs a @Convert

**Review by porritta — APPROVED** [review]
LGTM

**Pull request merged**